File: canit-api-client.php
<?php
/**
 * PHP API client library.  Simple wrapper around curl functions
 *
 * If you are planning on calling the CanIt API from PHP, you should
 * use this library.  It's MUCH easier than writing your own CURL calls.
 *
 * SYNOPSIS:
 *
 * require_once("canit-api-client.php");
 *
 *
 * CREATION: Create a CanItAPIClient object:
 *
 * $c = new CanItAPIClient("http://server.example.com/canit/api/2.0");
 *
 * LOGIN:
 *
 * $success = $c->login('username', 'password');
 * if ($success) {
 *     print "Successful login.\n";
 * } else {
 *     print "Login failed: " . $c->get_last_error() . "\n";
 * }
 *
 * REQUESTS:
 *
 * Do a GET request.  Example: Get a list of users.  Return value
 * is an array of arrays; each subarray is an associative array of users
 *
 * $users = $c->do_get('realm/@@/users');
 * if (!$c->succeeded()) {
 *     print "GET request failed: " . $c->get_last_error() . "\n";
 * } else {
 *     print "Got users:\n";
 *     print_r($users);
 * }
 *
 * Do a PUT request.  Example: Create a realm.
 *
 * $c->do_put('realm/foobar', array('description' => 'The new realm'));
 * if (!$c->succeeded()) {
 *     print "PUT request failed: " . $c->get_last_error() . "\n";
 * } else {
 *     print "Created realm foobar.\n";
 * }
 *
 * Do a POST request.  Example: Update a realm.
 *
 * $c->do_post('realm/foobar', array('description' => 'Different description'));
 * if (!$c->succeeded()) {
 *     print "POST request failed: " . $c->get_last_error() . "\n";
 * } else {
 *     print "Updated realm foobar.\n";
 * }
 *
 * Do a DELETE request.  Example: Delete a stream
 *
 * $c->do_delete('realm/@@/stream/wookie');
 * if (!$c->succeeded()) {
 *     print "DELETE request failed: " . $c->get_last_error() . "\n";
 * } else {
 *     print "Deleted stream wookie.\n";
 * }
 *
 * LOGOUT:
 *
 * $c->logout();
 */


class CanItAPIClient {
    /**
     * CanItAPIClient Constructor
     *
     * @param string $url The base URL of the API server.  Eg: http://127.0.0.1/canit/api/2.0
     */
    function CanItAPIClient ($url) {
	# Remove trailing slashes from $url
	$url = rtrim($url, '/');

	$url = $url . '/json/';

	$this->url = $url;
	$this->is_error = 0;
	$this->last_error = '';
	$this->curl_response = '';
	$this->curl_content = '';
	$this->curl_headers = '';
	$this->curl_content_type = '';
	$this->cookie = '';
    }

    /**
     * Get the last API-related error message
     * @return string
     */
    function get_last_error () {
	return $this->last_error;
    }

    /**
     * Returns true if the last API call failed,
     * false if it succeeded
     * DEPRECATED: You should use the succeeded() function instead
     */
    function is_error() {
	return $this->is_error;
    }

    /**
     * Returns true if the last API call succeeded,
     * false if it failed
     */
    function succeeded() {
	return ! $this->is_error;
    }

    /**
     * Log in to the API
     * @param string $username API username
     * @param string $password API password
     * @return boolean True on successful login; false otherwise
     */
    function login($username, $password) {
	$this->do_post('login', array('user' => $username,
				      'password' => $password));
	if ($this->is_error) {
	    return false;
	}

	# Set our cookie
	foreach ($this->curl_headers as $header) {
	    if (strpos($header, ':') !== FALSE) {
		list($name, $val) = explode(':', $header, 2);
	    } else {
		$name = $header;
		$val = '';
	    }
	    $name = strtolower(trim($name));
	    $val = trim($val);
	    if ($name == 'set-cookie') {
		if ($this->cookie != '') {
		    $this->cookie .= '; ';
		}
		$this->cookie .= $val;
	    }
	}
	return true;
    }

    /**
     * Log out of the API
     */
    function logout() {
	$this->do_get("logout");
	$this->cookie = '';
    }

    /**
     * Do a GET request
     * @param string $rel_url The relative URL.  That is everything
     *                        AFTER the /canit/api/2.0/ part of the full
     *                        URL.
     * @param array $params   Search array converted to ?key1=val1&key2=val2...
     * @return NULL on failure, a PHP data structure on success.
     */
    function do_get ($rel_url, $params = null) {
	# If $rel_url begins with a slash, remove it
	$rel_url = ltrim($rel_url, '/');

	$full_url = $this->url . $rel_url;
	if (is_array($params)) {
	    $first_time = 1;
	    foreach ($params as $key => $val) {
		if ($first_time) {
		    $full_url .= '?';
		    $first_time = 0;
		} else {
		    $full_url .= '&';
		}
		$full_url .= urlencode($key) . '=' . urlencode($val);
	    }
	}
	$ch = curl_init();
	$this->curl_call($full_url, $ch);
	curl_close($ch);
	return $this->deserialize_curl_data();
    }

    /**
     * Do a PUT request
     * @param string $rel_url The relative URL.  That is everything
     *                        AFTER the /canit/api/2.0/ part of the full
     *                        URL.
     * @param array $put_data An associative array of key/value pairs.
     * @return void Nothing useful; check $this->is_error() to test success
     */
    function do_put ($rel_url, $put_data) {
	# If $rel_url begins with a slash, remove it
	$rel_url = ltrim($rel_url, '/');

	$full_url = $this->url . $rel_url;
	$ch = curl_init();
	curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
	$encoded = json_encode($put_data);
	curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Length: ' . strlen($encoded)));
	curl_setopt($ch, CURLOPT_POSTFIELDS, $encoded);
	$this->curl_call($full_url, $ch);
	curl_close($ch);
	return $this->deserialize_curl_data();
    }

    /**
     * Do a DELETE request
     * @param string $rel_url The relative URL.  That is everything
     *                        AFTER the /canit/api/2.0/ part of the full
     *                        URL.
     * @return void Nothing useful; check $this->is_error() to test success
     */
    function do_delete ($rel_url) {
	# If $rel_url begins with a slash, remove it
	$rel_url = ltrim($rel_url, '/');

	$full_url = $this->url . $rel_url;
	$ch = curl_init();
	curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'DELETE');
	$this->curl_call($full_url, $ch);
	curl_close($ch);
	return NULL;
    }

    /**
     * Do a POST request
     * @param string $rel_url The relative URL.  That is everything
     *                        AFTER the /canit/api/2.0/ part of the full
     *                        URL.
     * @param array $put_data An associative array of key/value pairs.
     * @return void Nothing useful; check $this->is_error() to test success
     */
    function do_post ($rel_url, $post_data) {
	# If $rel_url begins with a slash, remove it
	$rel_url = ltrim($rel_url, '/');

	$full_url = $this->url . $rel_url;
	$ch = curl_init();
	curl_setopt($ch, CURLOPT_POST, true);
	curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
	$this->curl_call($full_url, $ch);
	curl_close($ch);
	return $this->deserialize_curl_data();
    }

    /** ==== END OF PUBLIC FUNCTIONS.  REMAINING FUNCTIONS ARE PRIVATE;
     ** ==== DO NOT CALL THEM DIRECTLY
     **/

    function curl_call($url, $ch) {
	curl_setopt($ch, CURLOPT_URL, $url);
	curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
	curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);
	curl_setopt($ch, CURLOPT_MAXREDIRS, 5);
	curl_setopt($ch, CURLOPT_TIMEOUT, 30);
	curl_setopt($ch, CURLOPT_HEADER, true);
	curl_setopt($ch, CURLOPT_FORBID_REUSE, true);
	if ($this->cookie != '') {
	    curl_setopt($ch, CURLOPT_COOKIE, $this->cookie);
	}
	curl_setopt($ch, CURLOPT_HTTPHEADER, array('Expect:', 'Accept: application/json'));
	$result = curl_exec($ch);
	if ($result === false) {
	    $this->is_error = 1;
	    $this->last_error = curl_error($ch);
	} else {
	    $arr = explode("\r\n\r\n", $result, 3);
	    $this->curl_headers = explode("\r\n", $arr[0]);
	    $this->curl_content = $arr[1];
	    $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
	    if ($code >= 200 && $code <= 299) {
		$this->curl_content_type = curl_getinfo($ch, CURLINFO_CONTENT_TYPE);
		$this->is_error = 0;
		$this->last_error = '';
	    } elseif ($code >= 400 && $code <= 599) {
		$this->is_error = 1;
		$this->set_error_from_result($this->curl_content);
	    } else {
		$this->is_error = 1;
		$this->last_error = "Unknown HTTP response $code";
	    }
	}
    }

    function set_error_from_result ($result) {
	# Special case: Login failures always come back in YAML
	# in old/buggy versions of API... sigh.
	if (substr($result, 0, 11) == "---\nerror: ") {
	    $this->last_error = substr($result, 11);
	    return;
	}
	$data = json_decode($result, true);
	if (!is_array($data)) {
	    $this->last_error = "Unknown error: $data";
	} elseif (array_key_exists('error', $data)) {
	    $this->last_error = $data['error'];
	} else {
	    $this->last_error = 'Unknown error';
	}
    }

    function deserialize_curl_data () {
	if ($this->is_error) return NULL;
	if ($this->curl_content_type == 'message/rfc822') {
	    return array('message' => $this->curl_content);
	}

	return json_decode($this->curl_content, true);
    }

}

?>

File: CanItAPIClient.py
import pycurl
import json
import urllib
import StringIO

# Python API client library.  Simple wrapper around curl functions
#
#
# SYNOPSIS:
#
# import CanItAPIClient
#
#
# CREATION: Create a CanItAPIClient object:
#
# c = CanItAPIClient.CanItAPIClient("http://server.example.com/canit/api/2.0")
#
# LOGIN:
#
# success = c.login('username', 'password')
# if success:
#     print "Successful login"
# else:
#     print "Login failed: " + c.get_last_error()
#
# REQUESTS:
#
# Do a GET request.  Example: Get a list of users.  Return value
# is an array of dictionaries.
#
# users = c.do_get('realm/@@/users')
# if c.succeeded():
#     print "Got users"
#     print users
# else:
#     print "GET request failed: " + c.get_last_error()
#
# Do a PUT request.  Example: Create a realm.
#
# c.do_put('realm/foobar', {'description' : 'The new realm'})
# if c.succeeded():
#     print "Created realm foobar."
# else:
#     print "PUT request failed: " + c.get_last_error()
#
# Do a POST request.  Example: Update a realm.
#
# c.do_post('realm/foobar', {'description' : 'Different description'})
# if c.succeeded():
#     print "Updated realm foobar."
# else:
#     print "POST request failed: " + c.get_last_error()
#
# Do a DELETE request.  Example: Delete a stream
#
# c.do_delete('realm/@@/stream/wookie')
# if c.succeeded():
#     print "Deleted stream wookie.\n"
# else:
#     print "DELETE request failed: "  + c.get_last_error()
#
# LOGOUT:
#
# c.logout()

class CanItAPIClient:
    """A simple client-side library for accessing the CanIt API"""
    def __init__(self, url):
        # Remove trailing slashes from url
        url = url.rstrip('/')
        url = url + '/json/'
        self.url = url
        self.is_error = 0
        self.last_error = ''
        self.curl_response = ''
        self.curl_content = ''
        self.curl_headers = ''
        self.cookie = ''
        pycurl.global_init(pycurl.GLOBAL_ALL)

    def get_last_error(self):
        """Returns the last CURL error if most recent call failed."""
        return self.last_error

    def succeeded(self):
        """Returns true if most recent API call succeeded; false otherwise."""
        return not self.is_error

    def login(self, username, password):
        """Log in to the API."""
        self.do_post('login', [('user', username), ('password', password)])
        if self.is_error:
            return False
        for header in self.curl_headers:
            lst = header.split(':', 2)
            if len(lst) < 2:
                continue
            name = lst[0].lower()
            val = lst[1].strip()
            if name == 'set-cookie':
                if self.cookie != '':
                    self.cookie = self.cookie + '; '
                self.cookie = self.cookie + val
        return True

    def logout(self):
        """Log out of the API and release cookie."""
        self.do_get('logout')
        self.cookie = ''
        return True

    def do_get(self, rel_url):
        """Do a GET request against the API server."""
	# If rel_url begins with a slash, remove it
        rel_url = rel_url.lstrip('/')
        full_url = self.url + rel_url
        c = pycurl.Curl()
        self.curl_call(full_url, c)
        c.close()
        return self.deserialize_curl_data()

    def do_put(self, rel_url, put_data):
        """Do a PUT request against the API server.  put_data should
        be a dictionary."""
	# If rel_url begins with a slash, remove it
        rel_url = rel_url.lstrip('/')
        full_url = self.url + rel_url
        c = pycurl.Curl()
        c.setopt(pycurl.CUSTOMREQUEST, 'PUT')

        # We don't know which version of the JSON API
        # we have... sigh...
        try:
            encoded = json.dumps(put_data)
        except AttributeError:
            encoded = json.write(put_data)

        c.setopt(pycurl.POSTFIELDS, encoded)
        c.setopt(pycurl.HTTPHEADER, ['Content-Type: application/json',
                                     'Content-Length: ' + str(len(encoded))])
        self.curl_call(full_url, c)
        c.close()
        return self.deserialize_curl_data()


    def do_delete(self, rel_url):
        """Do a DELETE request against the API server."""
	# If rel_url begins with a slash, remove it
        rel_url = rel_url.lstrip('/')
        full_url = self.url + rel_url
        c = pycurl.Curl()
        c.setopt(pycurl.CUSTOMREQUEST, 'DELETE')
        self.curl_call(full_url, c)
        c.close()
        return None

    def do_post(self, rel_url, post_data):
        """Do a POST request against the API server.  post_data should
        be a dictionary."""
	# If rel_url begins with a slash, remove it
        rel_url = rel_url.lstrip('/')
        full_url = self.url + rel_url
        c = pycurl.Curl()
        c.setopt(pycurl.POST, True)
        c.setopt(pycurl.POSTFIELDS, urllib.urlencode(post_data))
        self.curl_call(full_url, c)
        c.close()
        return self.deserialize_curl_data()

#==== END OF PUBLIC FUNCTIONS.  REMAINING FUNCTIONS ARE PRIVATE;
#==== DO NOT CALL THEM DIRECTLY

    def curl_call(self, url, c):
        c.setopt(pycurl.URL, url);
        c.setopt(pycurl.FOLLOWLOCATION, 1);
        c.setopt(pycurl.CONNECTTIMEOUT, 10);
        c.setopt(pycurl.MAXREDIRS, 5);
        c.setopt(pycurl.TIMEOUT, 30);
        c.setopt(pycurl.HEADER, 1);
        c.setopt(pycurl.FORBID_REUSE, 1)

	if (self.cookie != ''):
            c.setopt(pycurl.COOKIE, self.cookie)

	c.setopt(pycurl.HTTPHEADER, ['Expect:', 'Accept: application/json'])
        ans = StringIO.StringIO()
        c.setopt(pycurl.WRITEFUNCTION, ans.write)
        try:
            c.perform()
        except:
            self.is_error = 1
            self.last_error = c.errstr()
            return None

        lst = ans.getvalue().split("\r\n\r\n", 3)
        self.curl_headers = lst[0].split("\r\n")
        self.curl_content = lst[1]

        code = c.getinfo(pycurl.HTTP_CODE)
        if code >= 200 and code <= 299:
            self.is_error = 0
            self.last_error = ''
        elif code >= 400 and code <= 599:
            self.is_error = 1
            self.set_error_from_result(self.curl_content, code)
        else:
            self.is_error = 1
            self.last_error = 'Unknown HTTP response ' + str(code)


    def set_error_from_result(self, result, code):
        code = str(code)
        if (result == ''):
            self.last_error = 'Unknown error: HTTP Code ' + code
            return
        try:
            data = json.loads(result)
        except AttributeError:
            try:
                data = json.read(result)
            except:
                self.last_error = 'Unknown error: HTTP Code ' + code
                return
        except:
            self.last_error = 'Unknown error: HTTP Code ' + code
            return

        if not isinstance(data, dict):
            self.last_error = 'Unknown error: HTTP Code ' + code
            return

        if data.has_key('error'):
            self.last_error = data['error']
            return

        self.last_error = 'Unknown error: HTTP Code ' + code

    def deserialize_curl_data(self):
        if (self.curl_content == ''):
            return None

        if (self.is_error):
            return None

        try:
            ans = json.loads(self.curl_content)
        except AttributeError:
            ans = json.read(self.curl_content)

        return ans
